File: sangsoo/0x18 Graph/bfsdfs.py
from collections import deque

adj = [[] for _ in range(10)]
vis = [0 for _ in range(10)]

# BFS 연결 그래프에서 순회, 시작 노드 v
def bfs_traverse(v:int)->None:
    q = deque()
    q.append(v)
    vis[v] = 1
    while(q):
        cur = q.popleft()
        print(cur, end = ' ')
        for nxt in adj[cur]:
            if(vis[nxt]): continue
            vis[nxt] = 1
            q.append(nxt)

# BFS 연결그래프에서 v번 노드와 정점과의 거리

dist = [-1 for _ in range(10)]
def bfs_dist(v: int)->None:
    q = deque()
    q.append(v)
    dist[v] = 0
    while(q):
        cur = q.popleft()
        for nxt in adj[cur]:
            if(dist[nxt] != -1): continue
            q.append(nxt)
            dist[nxt] = dist[cur]+1

# BFS 연결 그래프가 아닐 때 순회
adj = [[] for _ in range(10)]
vis = [0 for _ in range(10)]
v = 9 # 정점의 수

def bfs_traverse2():
    q = deque()
    for i in range(1, 10):
        if (vis[iW]): continue
        q.append(i)
        vis[i] = 1
        while(q):
            cur = q.popleft()
            print(cur, end = ' ')
            for nxt in adj[cur]:
                if(vis[cur]): continue
                q.append(nxt)
                vis[nxt] = 1

# DFS 연결 그래프에서 순회, 비재귀 (재귀와 같게 출력)
adj = [[] for _ in range(10)]
vis = [0 for _ in range(10)]

def dfs(v:int)->None: 
    s = deque()
    s.append(v)
    while(s):
        cur = s.pop()
        if(vis[cur]): continue
        vis[cur] = 1
        print(cur, end = ' ')
        for i in range(len(adj[cur])):
            # 스택의 특성상 정점을 역순으로 넣어야함
            nxt = adj[cur][len(adj[cur])-1-i]
            if(vis[nxt]): continue
            s.append(nxt)

# DFS 연결 그래프에서 순회, 재귀
adj = [[] for _ in range(10)]
vis = [0 for _ in range(10)]

def dfs(v:int)->None:
    vis[v] =1
    print(v, end = ' ')
    for nxt in adj[v]:
        if (vis[nxt]): continue
        dfs(nxt)

File: sangsoo/0x18 Graph/bfsdfs.cpp
#include <vector>
#include <iostream>
#include <algorithm>
#include <queue>
#include <stack>

using namespace std;

vector<int> adj[10];
int vis[10];
// BFS 연결 그래프에서 순회, 시작 노드 v
void bfs_traverse(int v){
    queue<int> q;
    q.push(v);
    vis[v] = 1;
    while(!q.empty()){
        int cur = q.front(); q.pop();
        cout << cur << ' ';
        for(auto nxt : adj[cur]){
            if(vis[nxt]) continue;
            vis[nxt] = 1;
            q.push(nxt);
        }
    }
}

// BFS 연결 그래프에서 V번과 정점과의 거리
vector<int> adj[10];
int dist[10];
void bfs_dist(int v){
    fill(dist, dist+10, -1);
    queue<int> q;
    q.push(v);
    dist[v] = 0;
    while(!q.empty()){
        int cur = q.front(); q.pop();
        for(auto nxt : adj[cur]){
            if(dist[nxt] != -1) continue;
            q.push(nxt);
            dist[nxt] = dist[cur]+1;
        }
    }
}

// BFS 연결그래프가 아닐 때 순회
vector<int> adj[10];
bool vis[10];
int v = 9; // 정점의 수
void bfs_traverse2(){
    queue<int> q;
    for(int i = 1; i <= v; i++){
        if(vis[i]) continue;
        q.push(i);
        vis[i] = true;
        while(!q.empty()){
            int cur = q.front(); q.pop();
            cout << cur << ' ';
            for(auto nxt : adj[cur]){
                if(vis[nxt]) continue;
                q.push(nxt);
                vis[nxt] = 1;
            }
        }
    }
}

// DFS 연결그래프에서 순회, 비재귀 ver1 시작노드 v;
vector<int> adj[10];
int vis[10];
void dfs(){
    stack<int> s;
    s.push(v);
    vis[v] = 1;
    while(!s.empty()){
        int cur = s.top();
        s.pop();
        cout << cur << ' ';
        for(auto nxt : adj[cur]){
            if(vis[nxt]) continue;
            s.push(nxt);
            vis[nxt] = 1;
        }
    }
}

// DFS 연결그래프에서 순회, 비재귀 ver2 (재귀와 같게 출력), 시작노드 v
vector<int> adj[10];
int vis[10];
void dfs(int v){
    stack<int> s;
    s.push(v);
    // vis[1] = 1; 삭제
    while(!s.empty()){
        int cur = s.top();
        s.pop();
        if(vis[cur]) continue;
        vis[cur] = 1; // 추가
        cout << cur << ' ';
        for(int i = 0; i < adj[cur].size(); i++){
            // 스택의 특성상 정점을 역순으로 넣어야함
            int nxt = adj[cur][adj[cur].size()-1-i];
            if(vis[nxt]) continue;
            s.push(nxt);
            // vis[nxt] = 1; 삭제
        }
    }
}


// DFS 연결 그래프에서 순회, 재귀

vector<int> adj[10];
int vis[10];
void dfs(int cur){
    vis[cur] = 1;
    cout << cur << ' ';
    for(auto nxt : adj[cur]){
        if(vis[nxt]) continue;
        dfs(nxt);
    }

}